MAID-2957: Implement the proposed new design

octol opened the pull request:
Initial work on implementing the proposed new design. Very rough work so far, so be kind!

Usually I'd probably have squashed some of these commits, but I kept them granular for now.

Missing (that I'm aware of) are primarily the factory functions as well as serialisation stuff.

povilasb commented on src/lib.rs:
```diff
@@ -71,8 +71,42 @@ use serde::{de::DeserializeOwned, Serialize};
+    /// Decryption can fail because of invalid keys, invalid data, or deserialisation failures.
+    #[derive(Debug)]
+    pub enum Error {
+        /// Occurs when serialisation or deserialisation fails.
```
I know this is mostly just copy paste from the bottom, but still there's inconsistent language use:
```
Occurs when serialisation or deserialisation fails.

vs


This error is returned if encryption or decryption fail.
```
I'm talking about fails vs fail :)

octol replied:
Fixed in the last set of commits

povilasb commented on src/lib.rs:
```diff
@@ -92,76 +126,23 @@ pub fn hash(data: &[u8]) -> [u8; 32] {
+    /// Construct public key from bytes. Useful when it was serialized before.
+    pub fn from_bytes(public_key: [u8; PUBLIC_ENCRYPT_KEY_BYTES]) -> Self {
+        Self {
+            encrypt: box_::PublicKey::from_slice(&public_key).unwrap(),
```
We use `unwrap!` macro from https://crates.io/crates/unwrap where applicable - it gives us better output on panic.

octol replied:
Currently in lib.rs that crate seems to be hidden behind `#[cfg(any(test, feature = "mock"))]`. So is it ok to just remove that line?

povilasb replied:
yep :+1:

povilasb commented on src/lib.rs:
```diff
@@ -92,76 +126,23 @@ pub fn hash(data: &[u8]) -> [u8; 32] {
-    #[cfg(not(feature = "mock"))]
-    pub fn public_sign_key(&self) -> PublicSignKey {
-        self.sign.0
+    /// For anyone who wants to store public key.
```
"to store **the** public key`?

povilasb commented on src/lib.rs:
```diff
@@ -190,36 +171,42 @@ impl PublicKeys {
+/// Store secret key in `Arc` to discourage cloning it. Makes it more secure by not duplicating
+/// encryption keys in memory.
+/// Use `encrypt_key_pair()` to generate public and private keys pair.
+// NOTE, serde by default doesn't allow to serialize `Arc`. We have to enable this feature
```
It was meant as internal comment, hence I would remove it :)

povilasb commented on src/lib.rs:
```diff
@@ -190,36 +171,42 @@ impl PublicKeys {
-    pub fn verify_detached(&self, signature: &Signature, data: &[u8]) -> bool {
-        sign::verify_detached(&signature.signature, data, &self.sign)
-    }
+/// Store secret key in `Arc` to discourage cloning it. Makes it more secure by not duplicating
```
I would reword it to smth like "Reference counted secret encryption key used to decrypt data previously encrypted with `PublicEncryptKey`"

povilasb commented on src/lib.rs:
```diff
@@ -190,36 +171,42 @@ impl PublicKeys {
+    }
+
+    /// For anyone who wants to store secret key.
+    pub fn as_bytes(self) -> [u8; SECRET_ENCRYPT_KEY_BYTES] {
```
Reword to smth like "Get the inner secret key representation."?

povilasb commented on src/lib.rs:
```diff
@@ -244,13 +237,86 @@ impl SecretKeys {
+    (pub_enc_key, sec_enc_key)
+}
+
+/// Public signing key
```
Docs don't say more than the struct name itself. How about adding a sentence or two what's this signing key about? There's some info at https://download.libsodium.org/doc/public-key_cryptography/public-key_signatures.html

octol commented:
Added a few more commits, hopefully it should build and test correctly now.

Still working on reformulating some of the docs. Also need to redo the mocked `from_slice` functions as they currently just set zeros. Would be good to have a little unit test stats as well. 

And not to forget, switch to the `unwrap!` macro.

povilasb commented on src/mock_crypto.rs:
```diff
@@ -71,6 +71,13 @@ pub(crate) mod crypto_impl {
 
+            impl PublicKey {
+                // TODO: just to make it build
+                pub(crate) fn from_slice(_bs: &[u8]) -> Option<Self> {
```
hm, do we really need `from_slice()`, if we have `from_bytes()`? :)

octol replied:
I think these `from_slice` functions are mocking `rust_sodium` though?

povilasb replied:
Yeah, you're right. I missed the file name, sorry :)

octol committed:
refactor/all: expose crypto primitives

Different MaidSafe tools need different crypto primitives. E.g. Crust
only uses encryption public/private keys, Routing uses encryption and
signing public/private keys, etc. Currently in some cases safe_crypto is
unusable cause it doesn't expose some internal data like private keys,
etc.

We try to solve this by having 1-to-1 mappings with lower level crypto
types from rust_sodium etc. E.g.

- PublicSignKey
- SecretSignKey
- PublicEncryptKey
- SecretEncryptKey
- SymmetricKey
- Signature

[MAID-2957]

povilasb reviewed (dismissed):
Looks good to me

Fraser999 commented on src/lib.rs:
```diff
@@ -71,8 +71,42 @@ use serde::{de::DeserializeOwned, Serialize};
+        }
+        /// Occurs in case of an error during initialisation.
+        InitError(e: i32) {
+            display("error while initialising random generator")
```
I think this should be "error while initialising safe_crypto", since it's the whole lib being initialised.  Same applies to the doc comment for the `init()` function just below here.

Fraser999 commented on src/lib.rs:
```diff
@@ -71,8 +71,42 @@ use serde::{de::DeserializeOwned, Serialize};
+}
 
 /// Initialise random number generator for the key generation functions.
 pub fn init() -> Result<(), ()> {
```
I think it'd be better to return an actual `Error` here rather than the unit type - so e.g.
```rust
pub fn init() -> Result<(), Error> {
    crypto_impl::init().map_err(|_| Error::InitError(-1))
}
```

Fraser999 commented on src/lib.rs:
```diff
@@ -92,76 +126,23 @@ pub fn hash(data: &[u8]) -> [u8; 32] {
 
-/// Represents a set of public keys, consisting of a public signature key and a public
-/// encryption key.
+/// Use `gen_encrypt_keypair()` to generate a public and private key pair.
```
I think this doc comment should first state what `PublicEncryptKey` is rather than how to create one.  Also, we're using the term "secret" rather than "private" when describing crypto keys now.

Fraser999 commented on src/lib.rs:
```diff
@@ -92,76 +126,23 @@ pub fn hash(data: &[u8]) -> [u8; 32] {
-    ciphertext: Vec<u8>,
-}
+impl PublicEncryptKey {
+    /// Construct public key from bytes. Useful when it was serialized before.
```
We try to stick to UK spellings rather than US wherever possible, so "serialised" rather than "serialized".

Fraser999 commented on src/lib.rs:
```diff
@@ -92,76 +126,23 @@ pub fn hash(data: &[u8]) -> [u8; 32] {
-    #[cfg(not(feature = "mock"))]
-    pub fn public_sign_key(&self) -> PublicSignKey {
-        self.sign.0
+    /// For anyone who wants to store the public key.
```
Rather than a specific use-case, it might be better to just state that this converts the `PublicEncryptKey` into the raw underlying bytes or something like that?

Also, [a Rust convention](https://rust-lang-nursery.github.io/rust-clippy/v0.0.207/index.html#wrong_self_convention) is that a method like this which consume `self` should be named `into_...`.  I just noticed that [clippy is actually failing on this PR](https://travis-ci.org/maidsafe/safe_crypto/jobs/415933111) because of these `as_bytes()` methods (but it didn't show as a failure on the CI run because of a clippy bug which is resolved in a later version of clippy).

octol replied:
Ah thanks, ok I've updated the names to `into_...`

Fraser999 commented on src/lib.rs:
```diff
@@ -190,36 +171,40 @@ impl PublicKeys {
     pub fn encrypt_anonymous_bytes(&self, plaintext: &[u8]) -> Vec<u8> {
```
I think `anonymously_encrypt()` and `anonymously_encrypt_bytes()` are clearer function names, but maybe confirm with @ustulation before making this change if you agree?

Fraser999 commented on src/lib.rs:
```diff
@@ -190,36 +174,40 @@ impl PublicKeys {
-    }
+/// Reference counted secret encryption key used to decrypt data previously encrypted with
+/// `PublicEncryptKey`.
+/// Use `gen_encrypt_keypair()` to generate a public and private key pair.
```
"private" → "secret"

Fraser999 commented on src/lib.rs:
```diff
@@ -190,36 +174,40 @@ impl PublicKeys {
-                sign: sign_sk,
-                encrypt: encrypt_sk,
+impl SecretEncryptKey {
+    /// Construct secret key from given bytes. Useful when secret key was serialized before.
```
"serialized" → "serialised"

Fraser999 commented on src/lib.rs:
```diff
@@ -244,13 +238,88 @@ impl SecretKeys {
+    }
+
+    /// Convert the `PublicSignKey` into the raw underlying bytes.
+    /// For anyone who wants to store the public signature key.
```
"signature" → "signing"

Fraser999 commented on src/lib.rs:
```diff
@@ -260,21 +329,41 @@ impl SecretKeys {
-        let precomputed = Arc::new(box_::precompute(&their_pk.encrypt, &self.inner.encrypt));
-        SharedSecretKey { precomputed }
-    }
+/// Construct random public/private signing key pair.
```
"public/private" → "public and secret"

Fraser999 commented on src/lib.rs:
```diff
@@ -260,21 +329,41 @@ impl SecretKeys {
-    fn default() -> Self {
-        Self::new()
-    }
+/// Precomputed shared secret key.
```
I would break this doc comment block into two paragraphs (i.e. add a line with just `///` immediately after the first comment line) in order to keep the summary text fairly brief.

Fraser999 commented on src/lib.rs:
```diff
@@ -343,13 +432,32 @@ impl SharedSecretKey {
+}
+
 impl Signature {
+    /// Construct the signature from bytes. Useful when it was serialized before.
```
"serialized" → "serialised"

Fraser999 commented on src/lib.rs:
```diff
@@ -426,41 +547,42 @@ impl Default for SymmetricKey {
+            f,
+            "{:02x}{:02x}{:02x}..",
+            &self.encrypt.0[0], &self.encrypt.0[1], &self.encrypt.0[2]
+        )?;
```
If you remove `?;` here, you don't need to return `Ok(())` below.  Same for other two `impl fmt::Display`s below.

Fraser999 commented on src/lib.rs:
```diff
@@ -260,21 +329,41 @@ impl SecretKeys {
 }
 
 impl SharedSecretKey {
+    /// Construct shared secret key from bytes. Useful when it was serialized before.
```
"serialized" → "serialised"
